cmd/kaspawallet/daemon/server/split_transaction.go

```diff
@@ -0,0 +1,258 @@
+		partiallySignedInput := transaction.PartiallySignedInputs[i]
+		selectedUTXOs = append(selectedUTXOs, &libkaspawallet.UTXO{
+			Outpoint: &transaction.Tx.Inputs[i].PreviousOutpoint,
+			UTXOEntry: utxo.NewUTXOEntry(
+				partiallySignedInput.PrevOutput.Value, partiallySignedInput.PrevOutput.ScriptPublicKey,
+				false, constants.UnacceptedDAAScore),
+			DerivationPath: partiallySignedInput.DerivationPath,
+		})
+
+		totalSompi += selectedUTXOs[i-startIndex].UTXOEntry.Amount()
+		totalSompi -= feePerInput
+	}
+	unsignedTransactionBytes, err := libkaspawallet.CreateUnsignedTransaction(s.keysFile.ExtendedPublicKeys,
+		s.keysFile.MinimumSignatures,
+		[]*libkaspawallet.Payment{{
+			Address: changeAddress,
+			Amount:  totalSompi,
+		}}, selectedUTXOs)
+	if err != nil {
+		return nil, err
+	}
+
+	return serialization.DeserializePartiallySignedTransaction(unsignedTransactionBytes)
+}
+
+func (s *server) estimateMassAfterSignatures(transaction *serialization.PartiallySignedTransaction) (uint64, error) {
+	transaction = transaction.Clone()
+	var signatureSize uint64
+	if s.keysFile.ECDSA {
+		signatureSize = secp256k1.SerializedECDSASignatureSize
+	} else {
+		signatureSize = secp256k1.SerializedSchnorrSignatureSize
+	}
+
+	for i, input := range transaction.PartiallySignedInputs {
+		for j, pubKeyPair := range input.PubKeySignaturePairs {
+			if uint32(j) >= s.keysFile.MinimumSignatures {
+				break
+			}
+			pubKeyPair.Signature = make([]byte, signatureSize+1) // +1 for SigHashType
+		}
+		transaction.Tx.Inputs[i].SigOpCount = byte(len(input.PubKeySignaturePairs))
+	}
+
+	transactionWithSignatures, err := libkaspawallet.ExtractTransactionDeserialized(transaction, s.keysFile.ECDSA)
+	if err != nil {
+		return 0, err
+	}
+
+	return s.txMassCalculator.CalculateTransactionMass(transactionWithSignatures), nil
+}
+
+func (s *server) moreUTXOsForMergeTransaction(alreadySelectedUTXOs []*libkaspawallet.UTXO, requiredAmount uint64) (
+	additionalUTXOs []*libkaspawallet.UTXO, totalValueAdded uint64, err error) {
+
+	dagInfo, err := s.rpcClient.GetBlockDAGInfo()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	for _, utxo := range s.utxosSortedByAmount {
+		if !isUTXOSpendable(utxo, dagInfo.VirtualDAAScore, s.params.BlockCoinbaseMaturity) {
+			continue
+		}
+		for _, alreadySelectedUTXO := range alreadySelectedUTXOs {
+			if alreadySelectedUTXO.Outpoint.Equal(utxo.Outpoint) {
+				// comparing outpoints should be sufficient as they are unique per utxo entry
+				continue
+			}
+		}
+		additionalUTXOs = append(additionalUTXOs, &libkaspawallet.UTXO{
+			Outpoint:       utxo.Outpoint,
+			UTXOEntry:      utxo.UTXOEntry,
+			DerivationPath: s.walletAddressPath(utxo.address)})
+		totalValueAdded += utxo.UTXOEntry.Amount() - feePerInput
+		if totalValueAdded >= requiredAmount {
+			break
+		}
+	}
+	if totalValueAdded < requiredAmount {
+		return nil, 0, errors.Errorf("Insufficient funds for merge transaction")
```

Comment:
Do we really want to deny a user from sending funds just because they don't have enough money for a merge transaction?

Reply:
What's the alternative?

Reply:
Not having funds for merge transaction is more-or-less the same as not having enough fees, just send a little bit less and you'll be fine.